docs/commerce/payment/extend_payment.md

```diff
@@ -0,0 +1,120 @@
+
+## Create custom payment method type
+
+If your application needs payment methods of other type than the default `offline` one, you must create custom payment method types. 
+
+!!! note "Gateway integration requirement"
+
+    [[= product_name =]] does not come with gateway redirects. Whether you are an integrator or the end customer, it is your responsibility to implement payment gateway integration.
+
+### Define custom payment method type
+
+Create a PHP definition of the payment method type.
+
+``` php
+[[= include_file('code_samples/front/shop/payment/src/PaymentMethodType/PayPal/PayPal.php') =]]
+```
+
+Make sure that `getName` returns a human readable name of the payment method type, the way you want it to appear on the list of available payment method types.
+
+Now, register the definition as a service:
+
+``` yaml
+[[= include_file('code_samples/front/shop/payment/config/packages/services.yaml', 0, 5) =]]
+
+```
+
+As an alternative, instead of creating a PHP library, you can use the Type Factory to define the payment method type in your YAML configuration file:
+
+``` yaml
+services:
+    app.payment.type.paypal: 
+        class: Ibexa\Contracts\Payment\PaymentMethod\Type\TypeInterface
+        factory: [ '@Ibexa\Contracts\Payment\PaymentMethod\Type\TypeFactoryInterface', 'createType' ]
+        arguments: 
+            $identifier: 'paypal' 
+            $name: 'PayPal' 
+        tags: 
+            - name: ibexa.payment.payment_method.type 
+              alias: paypal
+
+```
+
+At this point a custom payment method type should be visible in the user interface.
+
+### Create options form
+
+Create a corresponding form type:
+
+``` php
+[[= include_file('code_samples/front/shop/payment/src/Form/Type/PayPalOptionsType.php') =]]
+```
+
+Next, create a mapper that maps the information that the user inputs in the form into attribute definition.
+
+``` php
+[[= include_file('code_samples/front/shop/payment/src/PaymentMethodType/PayPal/OptionsFormMapper.php') =]]
+```
+
+Then, register the `OptionsFormMapper` interface as a service:
+
+``` yaml
+services:
+    App\Payment\PayPal\OptionsFormMapper:
+        tags:
+            -   name: ibexa.payment.payment_method.options.form_mapper
+                type: paypal
+```
+
+### Create options validator
+
+You might want to make sure that data provided by the user is validated. 
+To do that, create an options validator that checks user input against the constraints and dispatches an error when needed.
+
+``` php
+[[= include_file('code_samples/front/shop/payment/src/PaymentMethodType/PayPal/UrlOptionValidator.php') =]]
+```
+
+Then, register the validator as a service:
+
+``` yaml
+services:
+    App\Payment\PayPal\OptionsValidator:
+        tags:
+            -   name: ibexa.payment.payment_method.options.validator
+                type: paypal
+```
+
+### Restart application
+
+Restart the application and create a payment of the new type.
```

Comment:
What do you mean by "Restart application" ?

Reply:
It is a term we've used in other places of the doc, telling people to shut down and start the server again.